### ISSUE TITLE
Typesafe generic response types and expansions

### ISSUE THREAD

**Issue by Daynil** [patmood/pocketbase-typegen#49]
Thank you for this typegen!! Makes it much nicer to work with pocketbase and typescript.

I had a few thoughts with regard to possible enhancements of some of the generic types that are generated.

Currently in the Responses, `Texpand` is unknown, so when we create typed expansions, you could easily typo the key or pass in an inappropriate response.

For example, if we have 2 records `articles` and `votes`, where you can expand the article associated with each vote. Currently, we can do:

```typescript
export type VotesResponse<Texpand = unknown> = VotesRecord & BaseSystemFields<Texpand>;

const vote = pb.collection('votes').getOne<
	VotesResponse<{
		article: ArticlesResponse;
	}>
>('a1', {
	expand: 'article'
});
```

However, if you accidentally pass ArticlesRecord instead of ArticlesReponse, you don't get a type error. We know that a "Response" variant is returned by pocketbase from an expansion, so it would be helpful to be able to enforce the type.

To do this, we need a generic for a response type, which I think can be achieved with:

```typescript
export type CollectionRecord = CollectionRecords[keyof CollectionRecords];
export type CollectionResponse<
	RequestedRecord extends CollectionRecord,
	Texpand extends {
		[key: string]: CollectionRecord &
			BaseSystemFields;
	} = Record<string, never>
> = RequestedRecord & BaseSystemFields<Texpand>;

const vote = pb.collection('votes').getOne<
	CollectionResponse<VotesRecord, {
		article: ArticlesResponse;
	}>
>('a1', {
	expand: 'article'
});
```

Now we would get an error if we accidently typed ArticlesRecord. You could also use this to generate the "Response types" for each of the collections, which you currently use `Texpand = unknown` for:

```typescript
export type VotesResponse<Texpand = unknown> = VotesRecord & BaseSystemFields<Texpand>

// to
type VotesResponse<
	Texpand extends {
		[K in SingularCollectionsUnion<CollectionsUnion>]?: CollectionRecord &
			BaseSystemFields;
	}
> = VotesRecord & BaseSystemFields<Texpand>;
```

We could further enhance this by typing the key of the expand, however this component uses an assumption of the convention of using the singular version of the collection name as the expand field (as suggested in the docs. But if we use this convention, we can also make the key of the expand typesafe (I also turned your Collections enum into a string union as they provide type safety with less boilerplate when possible):

```typescript
export type CollectionsUnion = `${Collections}`;
// If collection plural ends in ies, singular is y, else just remove the s
export type SingularCollectionsUnion<T extends CollectionsUnion> =
	T extends `${infer Prefix}ies`
		? `${Prefix}y`
		: T extends `${infer Prefix}s`
		? Prefix
		: T;

export type CollectionResponse<
	RequestedRecord extends CollectionRecord,
	Texpand extends {
		[K in SingularCollectionsUnion<CollectionsUnion>]?: CollectionRecord &
			BaseSystemFields;
	} = Record<string, never>
	> = RequestedRecord & BaseSystemFields<Texpand>;

const vote = await pb.collection('votes').getOne<
	CollectionResponse<VotesRecord, {
		article: ArticlesResponse<never>;
	}>
>('a1', {
	expand: 'article'
});
```

Now, if we accidently type `articles: ArticlesReponse` we would also get an error (because it should be `article`). That's getting a bit verbose, so having a `CollectionResponse` for each collection would be cool.

In any case, just brainstorming, I was surprised how far I could get making Texpand more typesafe without even touching your originally generated types, so technically folks could just create a separate file with these customizations.

**Comment by patmood** [comment]
Hey @Daynil thanks for the really great (detailed) suggestions. I'll test this out and see how it looks.

**Comment by Daynil** [comment]
Quick update, just realized the `Texpand` generic was still accepting the `Record` version in the expansion. Apparently since we are using intersection types to create the response, we need to be more specific.

What I ended up doing to fix this was creating a type with all responses, then using that in the Texpand, which properly enforces we are using a `Response`.

It also seems to work recursively (for nested expansions). Only piece I'm not sure of is how this would work with passing in the optional additional metadata type like you do in your [advanced usage example](https://github.com/patmood/pocketbase-typegen#example-advanced-usage).

```typescript
export type UsersResponse<
	Texpand extends {
		[K in SingularCollectionsUnion]?: CollectionResponse;
	} = Record<string, never>
> = UsersRecord & BaseSystemFields<Texpand>;

export type VotesResponse<
	Texpand extends {
		[K in SingularCollectionsUnion]?: CollectionResponse;
	} = Record<string, never>
> = VotesRecord & BaseSystemFields<Texpand>;

export type ArticlesResponse<
	Texpand extends {
		[K in SingularCollectionsUnion]?: CollectionResponse;
	} = Record<string, never>
> = ArticlesRecord & BaseSystemFields<Texpand>;

export type CollectionResponses = {
	articles: ArticlesResponse<any>;
	votes: VotesResponse<any>;
};
export type CollectionResponse = CollectionResponses[keyof CollectionResponses];

// error
const vote = await pb.collection('votes').getOne<
	VotesResponse<{
		article: ArticlesRecord<{
			user: UsersRecord
		}>;
	}>
>('a1', {
	expand: 'article,article.user'
});

// ok
const vote = await pb.collection('votes').getOne<
	VotesResponse<{
		article: ArticlesResponse<{
			user: UsersResponse
		}>;
	}>
>('a1', {
	expand: 'article,article.user'
});
```

**Comment by patmood** [comment]
Yeah I also ran into issues with the ordering for collections with json fields, since those generics come first:
`Required type parameters may not follow optional type parameters.`

I would love to have the expand argument first, but unfortunately the expand arguments were added after the json field arguments so it would be a backward incompatible change.

Expanded relations are used less frequently than single fetches and if this was implemented, I think you would be required to pass in a some arg even when you weren't using expand field. 

Unless it can be optional I don't think the benefit from this outweighs the need to pass in `{}` every time.

I really appreciate the detail you've put into this and I think if we just add `CollectionResponses` to the generated fields, that would unblock you and others to implement this technique on top.
